Support for cancelling part movements between shards

nvartolomei opened the pull request:
Follow up to #17871. Won't be ready for merge for some time, here for early feedback.

Part of #13574
---
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

---

- [ ] Use different prefix for cloned part to avoid collision with detached parts when cloning replicas. clone_

nvartolomei committed:
Support for cancelling part movements between shards

nvartolomei commented:
@alesapin can you please go over the `TODO(nv)` left in the code and leave some early feedback please?

alesapin commented on src/Interpreters/InterpreterKillQueryQuery.cpp:
```diff
@@ -290,6 +290,73 @@ BlockIO InterpreterKillQueryQuery::execute()
+        MutableColumns res_columns = header.cloneEmptyColumns();
+        auto table_id = StorageID::createEmpty();
+        AccessRightsElements required_access_rights;
+//        TODO(nv): Implement ACLs.
```
Yes, we need to implement ACLs

alesapin commented on src/Parsers/ASTKillQueryQuery.h:
```diff
@@ -13,6 +13,11 @@ class ASTKillQueryQuery : public ASTQueryWithOutput, public ASTQueryWithOnCluste
     {
         Query,      /// KILL QUERY
         Mutation,   /// KILL MUTATION
+        // TODO(nv): THIS_KILL_TASK_TYPE naming is ugly.
```
Why is it ugly? There are not so many things we can kill. I don't think that somebody actually moves parts `TO DISK` or `TO VOLUME` by hand.

alesapin commented on src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp:
```diff
@@ -139,7 +145,9 @@ bool PartMovesBetweenShardsOrchestrator::step()
         tryLogCurrentException(log, __PRETTY_FUNCTION__);
 
         Entry entry_copy = entry_to_process.value();
+        // TODO(nv): Clear these up on successful transitions.
```
+1

alesapin commented on src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp:
```diff
@@ -284,6 +327,18 @@ void PartMovesBetweenShardsOrchestrator::stepEntry(const Entry & entry, zkutil::
+            if (entry.rollback)
+            {
+                Entry entry_copy = entry;
+                // TODO(nv): Do we want to cleanup fetched data on the destination?
```
I think background cleanup is Ok.

alesapin commented on src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp:
```diff
@@ -331,13 +386,40 @@ void PartMovesBetweenShardsOrchestrator::stepEntry(const Entry & entry, zkutil::
+        case EntryState::DESTINATION_DROP:
         {
-            std::this_thread::sleep_for(std::chrono::seconds(storage.getSettings()->part_moves_between_shards_delay_seconds));
+            // TODO(nv): Move validation outside of the transitions. This is needed for all of them.
```
validation?

alesapin commented on src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp:
```diff
@@ -401,7 +483,9 @@ void PartMovesBetweenShardsOrchestrator::stepEntry(const Entry & entry, zkutil::
             Entry entry_copy = entry;
-            entry_copy.state = EntryState::DONE;
+
+            // TOOD(nv): Very confusing.
```
:) Yes

alesapin commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -6812,6 +6812,55 @@ void StorageReplicatedMergeTree::movePartitionToShard(
+CancellationCode StorageReplicatedMergeTree::killPartMoveToShard(const UUID & task_uuid)
+{
+    // TODO(nv): Needs cleanup, better methods and maybe some encapsulation.
+    // TODO(nv): Do we need SYNC/ASYNC handling? What does it mean?
```
No, we don't, but it would be nice to have some result from the kill command.

nvartolomei committed:
Extract state transitions to simplify stepEntry

nvartolomei committed:
Remove unnecessary code

nvartolomei committed:
Move kill logic to orchestrator

nvartolomei committed:
Defensive checks for rollback

nvartolomei committed:
Use correct states for rollback transitions

nvartolomei committed:
Improve DESTINATION_ATTACH idempotency

nvartolomei committed:
Store destination part name, can be useful for debugging and for tests

nvartolomei committed:
Rollback attach and failed attach

nvartolomei committed:
Debug message from step()

And always initialize num_entries.

nvartolomei committed:
Collect outsanding todos

nvartolomei committed:
Implemeent non-blocking waits for log entry processing

nvartolomei committed:
Merge remote-tracking branch 'ch/master' into nv/parts-uuid-move-shard-kill

* ch/master: (467 commits)
  fixed
  Update src/Common/Throttler.cpp
  done
  Implement antlr parser for projections (#24245)
  more
  simplify test_backup_restore
  done
  Paranoid fix for background pool task decrement
  fix style AggregateFunctionWindowFunnel.cpp
  Fix build and better test
  fix
  update
  update
  Set allow_experimental_funnel_functions = false by default (not used before)
  Remove example usage of allow_experimental_funnel_functions from windowFunnel
  docs: update requests (to fix conflicts with urllib3)
  Update requirements.txt
  Add exponential smoothing to Throttler
  Reverted libunwind from upstream
  Touch to rebuild
  ...

nvartolomei committed:
Merge branch 'nv/non-blocking-wait-log-entry' into nv/parts-uuid-move-shard-kill

* nv/non-blocking-wait-log-entry:
  Implemeent non-blocking waits for log entry processing

nvartolomei committed:
Avoid blocking for too long when waiting for log entries to be processed

nvartolomei committed:
Issue at most one destination fetch; allow to wait on older tasks

nvartolomei committed:
Hold entries in a vector rather than in a map

Preserved task creation order.

nvartolomei committed:
Use TIMEOUT_EXCEEDED error code on log entry wait timeouts

nvartolomei committed:
Consistent log entry submit/wait pattern

nvartolomei committed:
Fix garbled json parsing error

nvartolomei committed:
Note about how to test

nvartolomei committed:
Very pedantic part movement test

nvartolomei committed:
DROP_RANGE on rollback must be persisted as well

nvartolomei commented:
At this point I'm quite happy with the tests and that it works.

cc @xjewer @alesapin please review.

There are 2 open questions that need attention. If we agree on these I think this will be ready to be merged. This is the best I could come with taking into account time constraints.

The questions are related to each other and are result of me resorting to a tight coupling between shards. As a long term solution this is not great but given the direction we took it should be acceptable.

The tight coupling is coordination/communication between source and destination shard via ZooKeeper log rather than via a standardized API. A good standardized API is the SQL interface and to make part movement work via this interface we would need to go back to the idea of providing lower level primitives like ADD/REMOVE FINGERPRINT #14860, `WRITE PART` (from source to destination) and a way to really make these idempotent.

Because communication via ZooKeeper log is really a simplex communication channel it is hard to implement idempotency for sending a log entry and then waiting for its status. Keeping in mind that we already decided to accept tight coupling in previous PRs I added even more of it. When a log entry is created in ZooKeeper we also store "a reference" (in current implementation we store a copy of the log entry contents, search for `log_entry_barrier_path`). This provides idempotency for state transitions. Later on retries due to timeouts or crashes we use the “reference” stored at `log_entry_barrier_path` to find the initial log entry and check its processing status. You will also notice that `waitForAllTableReplicasToProcessLogEntry` is “non-blocking” now, it can just check the status and return instantly.

This leads me to open question #1. There isn’t a well-defined mechanism for doing what I just described. That is: create a log entry and persisting the log entry reference atomically so that it can be used later to check the processing status of the log entry. There was one (which is used in current PR) but I removed it few days ago #24960. I’m not sure i want to revert it back as it is a bit fragile and given how we actually want to use it (write to contents to ZK, later deserialize contents, create new log_entry that can be passed to `Strings StorageReplicatedMergeTree::waitForAllReplicasToProcessLogEntry( const ReplicatedMergeTreeLogEntryData & entry…)`, then compare serialized entry contents with log entry contents one by one to find if it is still in the queue) I can foresee this “roundtrip” ser/de breaking quite easily. Especially when running replicas on slightly different versions.

https://github.com/ClickHouse/ClickHouse/pull/24960/files#diff-0f4308161ab179a68f01cabfce4dc9e6e4e5f9c1e6c46c37d5adaeaad74e9913L5395-L5401
and
https://github.com/ClickHouse/ClickHouse/blob/82b8d45cd71128378b334e0aa4abc756898e688d/src/Storages/StorageReplicatedMergeTree.cpp#L5378

ClickHouse assumes entry is processed if it is not found in the queue, this would lead to data being dropped on source before it is fetched on destination.

We need a more reliable mechanism for identifying already-submitted log entries. One mechanism I was thinking of was to use a ZK path to first generate “log entry ids” (like ZK lock or just an ephemeral sequence?) then add an “ID” member to ReplicatedMergeTreeLogEntryData and also store it (atomically) at log_entry_barrier_path. Later use this ID when looking up the original log entry.

https://github.com/ClickHouse/ClickHouse/blob/82b8d45cd71128378b334e0aa4abc756898e688d/src/Storages/MergeTree/ReplicatedMergeTreeLogEntry.h#L79-L83

Open question #2. DROP_RANGE format depends on `storage.format_version` and during rollback we need to put a DROP_RANGE log entry in the destination shard log.

https://github.com/ClickHouse/ClickHouse/pull/24585/files#diff-8fd328c262b513261281d553c8f453189c12ebb1b302baf86eabc45f27488b13R422-R425

How can we assert that source format version and destination are the same or compatible? A check in ZK would be acceptable, but I couldn’t find this stored anywhere and is instead inferred from parts on disk. I’m pretty sure I’m missing something because this needs to match between replicas and most likely it is stored there.

Wdyt?

alesapin reviewed (commented):
The first question is quite complex, but obviously, we have solved it for replicated merge tree, let me think.

nvartolomei commented:
Cont. #29043
